### PR TITLE
Syncer/Reconciler Refactor

### PR DESCRIPTION
**Pull request by patrick-ogrady**
### Problems
* Multiple people expressed interest in using the `syncer` and `reconciler` packages in their own applications, however, these packages were tightly coupled to the other packages in `rosetta-cli`.
* In the last few weeks, significant tech debt was created to support running validation from an arbitrary start index without needing to sync from scratch. This lead to tons of duplicate logic in both the `reconciler` and `syncer` in the `stateful_x` and `stateless_x` implementations.
* Without unified logic, 2 commands were created to access the stateful (`check:complete`) and stateless syncer (`check:quick`). Having 2 commands to check validity was extremely confusing for new users.
* For blockchains that had genesis allocations, it was necessary to generate a bootstrap balances file to set balances before beginning syncing. This was a cumbersome tool to build for many Node API implementers.

### Solutions
* Remove all `internal` dependencies from `syncer` and `reconciler` to prepare for a migration to `rosetta-sdk-go`. It is now possible to provide your own storage and communication implementation.
* Abstract logic in `syncer` and `reconciler` to work regardless of whether starting from `genesis` or some arbitrary height. This made the implementation much simpler and removed plenty of redundant code.
* Create new `check` command that determines how to prepare storage to sync from any index. If you start from genesis, it will perform a comprehensive check. If you start from an arbitrary height, limited checks will be run.
* Storage will automatically fetch the balance of an account at the parent block before processing a block containing the account if it is not found. This means it is only necessary to create a bootstrap accounts file if it is not possible to lookup balance by block.

### Future PR
* Move `syncer` and `reconciler` packages to `rosetta-sdk-go`
* Restore the inactive checking queue from a previous instantiation

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -3,14 +3,15 @@
 	watch-transactions watch-balances watch-reconciliations \
 	view-block-benchmarks view-account-benchmarks
 LICENCE_SCRIPT=addlicense -c "Coinbase, Inc." -l "apache" -v
+GO_INSTALL=GO111MODULE=off go get
```
Comment (zquestz) [inline Makefile:6]: We really shouldn't be turning this flag off... this is the go to dep management in Go and is on by default in all the latest supported releases.

https://dev.to/maelvls/why-is-go111module-everywhere-and-everything-about-go-modules-24k#-raw-go111module-endraw-with-go-113

File: Makefile [hunk 2]
```diff
@@ -3,14 +3,15 @@
...
 
 deps:
 	go get ./...
 	go get github.com/stretchr/testify
-	go get github.com/google/addlicense
-	go get github.com/segmentio/golines
-	go get github.com/mattn/goveralls
+	${GO_INSTALL} github.com/google/addlicense
+	${GO_INSTALL} github.com/segmentio/golines
+	${GO_INSTALL} github.com/mattn/goveralls
```
Comment (zquestz) [inline Makefile:14]: All the deps we need should be in go.mod imho. It won't get compiled in unless used in code, so it doesn't bloat the resulting binary.

File: internal/reconciler/reconciler.go [hunk 3]
```diff
@@ -16,25 +16,624 @@ package reconciler
...
+		// Account changed on this block
+		if skipAccount {
+			continue
+		}
+
+		// If account + currency not found, add with difference 0
+		balanceChanges = append(balanceChanges, &BalanceChange{
+			Account:    account.Account,
+			Currency:   account.Currency,
+			Difference: "0",
```
Comment (zquestz) [inline internal/reconciler/reconciler.go:242]: Shouldn't this just use zeroString?

File: Makefile (hunk 1 above)
Comment (patrick-ogrady) [reply Makefile]: Found a solution: `go run X` for all commands. This still loads the dependency in the `go.mod` and allows for running the command.

**Pull request merged**